### ISSUE TITLE
Build XA/XI/XM tests with `make -C runtime mcs-do-test-profiles test_profiles="monodroid"`

### ISSUE THREAD

**Issue by luhenry** [mono/mono#10900]
## Steps to Reproduce

1. 
2. 
3. 

## On which platforms did you notice this

[ ] macOS
[ ] Linux
[ ] Windows

**Version Used**:

## Stacktrace

```
Please paste the stack trace here if available.
```

**Comment by luhenry** [comment]
@marek-safar since you largely edited the issue to something quite different to the original issue, let me clarify what's actually needed.

`compile-tests` is not useful for the SDKs since we already have `make -C runtime mcs-do-test-profiles mcs-do-xunit-test-profiles test_profiles="monodroid"` (`monodroid` for the case of XA for example) which generate all the test assemblies for the needed profiles. The problem is the build is broken, thus it needs to be fixed.

As for `package-tests`, I do not want to have multiple targets for something that should already be self-contained: `make -C runtime mcs-do-test-profiles mcs-do-xunit-test-profiles test_profiles="monodroid"` already contains the assumption we want the tests but also their dependencies since the tests are otherwise unusable.

**Comment by marek-safar** [comment]
> compile-tests is not useful for the SDKs

It might not be useful for SDKs but it's useful for Helix and I am not sure we need to maintain two very similar systems. I am happy to make that change in SDKs to call different target (We could reuse existing `make test`).

> As for package-tests
I am unconvinced it's a good idea to copy all the dependencies all the time during compilation time. The normal devs workflow will work just fine using MONO_PATH and for all-tests compilation mode, we need to do it only once as well. Secondly, not all tests will end up in `mcs/class/lib/<profile>/tests` and many have additional non-embedded dependencies for all profiles.

I also unticked the first checkbox because that work has not been completed.

**Comment by akoeplinger** [comment]
First of all we need to have a consistent way to precompile all test assemblies, right now it's a mix of calling `test`, `tests`, `compile-tests`, etc. targets in various parts of the repo. My ideal world would be to run `make test` from the root of the repo and it'll descend into each folder to precompile the tests.

A `package-tests` target would be useful and I agree with Marek that we don't want to copy stuff around all the time on a normal dev inner loop.

To give a good example: The mcs compiler tests are a collection of a few thousand .cs files that need to be present next to the test runner. Copying those all the time before a test execution is a waste of time.

**Comment by luhenry** [comment]
I don't understand the argument of "copying all the time": this is what `make` gives you, update only targets that needs to be updated to avoid unnecessary rebuilding and/or copying. If you end up copying it all the time, then something is broken in your dependency chain.

If we put all BCL tests with their dependencies into `mcs/class/lib/<profile>/tests` (which is what I am proposing for `make -C runtime mcs-do-test-profiles` - ie `make -C mcs test-profiles` - to do), then we have both `compile-tests` and `package-tests` ideas in 1 command; command which we already rely on in other places of our build/infrastructure.

The use case for Helix and the SDKs are very similar, so whatever already works for the SDKs will work for Helix as well. Having to copy only `mcs/class/lib/<profile>` to have all of the BCL and its test suite would fix both our problems.

**Comment by luhenry** [comment]
Ok, sounds good to me then :) Thanks for the clarification @akoeplinger

**Comment by luhenry** [comment]
As pointed out earlier, the `make -C runtime test xunit-test test_profiles="..."` doesn't work: https://jenkins.mono-project.com/job/test-mono-pull-request-amd64-osx-products-sdks-android/5314/parsed_console/log_content.html#ERROR1

**Comment by marek-safar** [comment]
Correct, that work is still pending (the checkbox for that is still unticked above)

**Comment by mandel-macaque** [comment]
This issue is quite interesting for the work we are doing here: https://github.com/xamarin/xamarin-macios/pull/4914 

We are using the test assemblies and are adding support on xharness to run them.

(this is a way to give you info on our side and cc myself to the issue ;) )
